### PR TITLE
CART-89 cart: Cart multiprov changes

### PR DESCRIPTION
**Pull request by frostedcmos**
CART-89 multiprovider: multiprovider stage1

Stage1

Signed-off-by: Alexander Oganezov <alexander.a.oganezov@intel.com>

### CODE REVIEW

**Review by NiuYawei — APPROVED** [review]
LGTM, some style warnings need be addressed.

File: src/tests/ftest/cart/dual_provider_common.h
Comment (kjacque) [inline src/tests/ftest/cart/dual_provider_common.h:131]: Could this #if / #endif pair be removed now?

**Review by kjacque — COMMENTED** [review]
Nothing huge jumping out at me, but I would like to get the fixable checkpatch issues fixed before landing. Not sure if you want to get any more review from CART devs before landing.

File: src/include/cart/types.h
Comment (soumagne) [inline src/include/cart/types.h:117]: Just a detail but since you have a lot of routines named `crt_prov` already you could consider naming that `crt_prov_t` instead ? it would also align with libfabric naming. (same for `CRT_PROVIDER` -> `CRT_PROV` ?)

File: src/cart/crt_context.c
Comment (soumagne) [inline src/cart/crt_context.c:203]: it seems dangerous to me here to return the URI like that ? since you would no longer know after that point who "owns" the memory of the URI. You should probably strdup() it ? or return it as const so that nobody tries to free it? but I think the safest is to dup it.

Comment (frostedcmos) [reply src/cart/crt_context.c:203]: Agreed, good catch, changed to D_STRNDUP()

File: src/include/cart/types.h
Comment (frostedcmos) [reply src/include/cart/types.h:117]: Sure, changed to CRT_PROV

File: src/tests/ftest/cart/dual_provider_common.h
Comment (frostedcmos) [reply src/tests/ftest/cart/dual_provider_common.h:131]: yes indeed

File: src/include/cart/types.h
Comment (soumagne) [reply src/include/cart/types.h:117]: Is there a reason for hard-coding the enum value of each crt prov ?

**Pull request merged**
